Binned PR-related metrics

maximsch2 opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #95 .

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

Borda commented:
@maximsch2 how is it going here, ready for review?

maximsch2 commented:
@Borda, I need to fix up isort and do a short cleanup, will get to it some time this week.

Borda commented:
@maximsch2 how is the clean-up doing? ;]

maximsch2 commented:
@Borda , I've moved out the function I wanted to do, feel free to take a look now. Another thing I think I should probably do is to move out test data generation code into a central place, but actually curious about your feedback to see if we should just replace the main one with it? 

The motivation is to get a reasonably-looking model, so that metrics like recall@precision=0.8 can be computed as the random data never actually reaches that precision :)

pep8speaks commented:
Hello @maximsch2! Thanks for updating this PR. 



* In the file [`tests/classification/test_binned_precision_recall.py`](https://github.com/PyTorchLightning/metrics/blob/eb70c497a131ff9da751f9a34b439894ba6b74eb/tests/classification/test_binned_precision_recall.py):

> [Line 124:121](https://github.com/PyTorchLightning/metrics/blob/eb70c497a131ff9da751f9a34b439894ba6b74eb/tests/classification/test_binned_precision_recall.py#L124): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 120 characters)

maximsch2 commented:
@Borda , mind taking a look now?

Borda commented on tests/classification/inputs.py:
```diff
@@ -77,3 +77,29 @@
+
+
+# Generate plausible-looking inputs
+def generate_plausible_inputs_multilabel():
```
lets rather have all the constants as arguments with defaults

Borda commented on tests/classification/test_binned_precision_recall.py:
```diff
@@ -0,0 +1,162 @@
+from sklearn.metrics import precision_recall_curve as _sk_precision_recall_curve
+
+from tests.classification.inputs import (
+    _input_binary_prob,
+    _input_binary_prob_plausible,
+    _input_multilabel_prob,
+    _input_multilabel_prob_plausible,
```
```suggestion
    _input_binary_prob,
    _input_binary_prob_plausible as _input_binary_prob_ok,
    _input_multilabel_prob as _input_mlb_prob,
    _input_multilabel_prob_plausible as _input_mlb_prob_ok,
```
let's shorten these imports or if you have better shortcuts :]

Borda reviewed (commented):
it mainly requires docs work:
- [ ] ref to the metric definition
- [ ] argument/options desc
- [ ] examples

maximsch2 commented:
Sounds good, let me work on that. While I started I've noticed another clean-up opportunity. Also, what's up with circleci? I see bleu failing intermittently even though I'm not touching it

maximsch2 commented:
> it mainly requires docs work:
> 
> * [x]  ref to the metric definition
> * [x]  argument/options desc
> * [x]  examples

One tricky thing is bleu tests, I've added atol there to make them pass stably as they are pretty close anyway. Please take another look.

justusschock commented on tests/classification/test_precision_recall.py:
```diff
@@ -195,7 +195,7 @@ def test_no_support(metric_class, metric_fn):
 class TestPrecisionRecall(MetricTester):
 
-    @pytest.mark.parametrize("ddp", [False])
+    @pytest.mark.parametrize("ddp", [False, True])
```
@SkafteNicki @Borda do we have multi-gpu tests here to run ddp?

Borda replied:
yes, we run on the very same cluster as PL

Borda replied:
TM: https://github.com/PyTorchLightning/metrics/blob/935994c8d56a6c9b2c97db6b86dcd224907b55e4/azure-pipelines.yml#L26

PL: https://github.com/PyTorchLightning/pytorch-lightning/blob/aaccbeea2beb3b903428643d2ee520d2dd05b348/azure-pipelines.yml#L26

SkafteNicki commented on tests/classification/test_binned_precision_recall.py:
```diff
@@ -0,0 +1,163 @@
+    @pytest.mark.parametrize("dist_sync_on_step", [True, False])
+    @pytest.mark.parametrize("min_precision", [0.05, 0.1, 0.3, 0.5, 0.8, 0.95])
+    def test_binned_pr(self, preds, target, sk_metric, num_classes, ddp, dist_sync_on_step, min_precision):
+        self.atol = 0.02
```
Move outside function like here:
https://github.com/PyTorchLightning/metrics/blob/ec239e00ebfad96fc9a0499c7bcd0d74e88dbffa/tests/regression/test_ssim.py#L69-L70
```suggestion

```

SkafteNicki commented on docs/source/references/modules.rst:
```diff
@@ -126,6 +126,12 @@ AveragePrecision
 .. autoclass:: torchmetrics.AveragePrecision
     :noindex:
 
+BinnedAveragePrecision
```
insert in alphabetical order

SkafteNicki commented on docs/source/references/modules.rst:
```diff
@@ -198,6 +204,12 @@ PrecisionRecallCurve
 .. autoclass:: torchmetrics.PrecisionRecallCurve
     :noindex:
 
+BinnedPrecisionRecallCurve
```
insert in alphabetical order

SkafteNicki commented on docs/source/references/modules.rst:
```diff
@@ -198,6 +204,12 @@ PrecisionRecallCurve
+
+.. autoclass:: torchmetrics.BinnedPrecisionRecallCurve
+    :noindex:
+
```
seems like `BinnedRecallAtFixedPrecision` is missing

SkafteNicki reviewed (commented):
smaller comments

maximsch2 commented:
Addressed comments. On docs, is your preference to go for strict alphabetical order? I was putting the binned versions close to non-binned ones so that people will notice them together, but I can move the new one to the to the sorted place as well.

Borda commented:
> I was putting the binned versions close to non-binned ones so that people will notice them together, but I can move the new one to the to the sorted place as well.

that is also an option, I do not have strong preference here :]

Borda commented on torchmetrics/classification/binned_precision_recall.py:
```diff
@@ -0,0 +1,303 @@
+        tensor([0.0000, 0.5000, 1.0000])]
+    """
+
+    TPs: Tensor
+    FPs: Tensor
+    FNs: Tensor
+    thresholds: Tensor
```
is this a counter?
if I have one metric for train other for valid it would mix all together, right?

maximsch2 replied:
This is just type annotation, it doesn't define a new variable and we use instance variables so there won't be overlap.

The cause here is the confusing way of Python type annotation for instance variables:
```python
class A:
   x: int # instance variable
   y: ClassVar[int] # class/static variable
```

https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations

Borda replied:
ok, I see not it is just kind or shadow to make IDE happy, right? https://github.com/PyTorchLightning/metrics/pull/128/files#r612738983
well, I find it confusing... :/
other thoughts? @PyTorchLightning/core-metrics

maximsch2 replied:
Yeah, it's for type-checker, I do agree that given that we don't have that in other metrics anyway it's fine to remove from here as well.

Borda replied:
yes, so let's make this PR without it and then open separate PR just by introducing this typing for all metrics... ok?

Borda commented on torchmetrics/classification/binned_precision_recall.py:
```diff
@@ -0,0 +1,303 @@
+        thresholds = torch.linspace(0, 1.0, num_thresholds)
+        self.register_buffer("thresholds", thresholds)
+
+        for name in ("TPs", "FPs", "FNs"):
```
ok, see here...
